core/mem/allocators.odin

```diff
@@ -982,3 +984,152 @@ tracking_allocator_proc :: proc(allocator_data: rawptr, mode: Allocator_Mode,
+		allocator_data = data.backing.data,
+		mode = mode,
+		size = size,
+		alignment = alignment,
+		old_memory = old_memory,
+		old_size = old_size,
+		location = loc,
+	)
+
+	if err == nil {
+		switch mode {
+		case .Query_Features, .Query_Info, .Resize:
+			// Ignore
+		case .Alloc, .Alloc_Non_Zeroed:
+			data.counter += 1
+		case .Free:
+			data.counter -= 1
+		case .Free_All:
+			data.counter = 0
+		}
+	}
+	return
+}
+
+
+Profiling_Allocator_Mode_Info :: struct {
+	count: int,
+	time:  time.Duration,
+}
+
+// Allocator for profiling number of allocations and their duration.
+Profiling_Allocator :: struct {
+	backing: Allocator,
+	modes:   [Allocator_Mode]Profiling_Allocator_Mode_Info,
+}
+
+profiling_allocator_init :: proc(data: ^Profiling_Allocator, backing := context.allocator) {
+	data^ = {
+		backing = backing,
+	}
+}
+
+@(require_results)
+profiling_allocator :: proc(data: ^Profiling_Allocator) -> Allocator {
+	return {
+		procedure = profiling_allocator_proc,
+		data = data,
+	}
+}
+
+profiling_allocator_proc :: proc(allocator_data: rawptr, mode: Allocator_Mode,
+                                 size, alignment: int,
+                                 old_memory: rawptr, old_size: int, loc := #caller_location) -> (result: []byte, err: Allocator_Error) {
+	data := cast(^Profiling_Allocator)allocator_data
+
+	start_time := time.now()
+	result, err = data.backing.procedure(
+		allocator_data = data.backing.data,
+		mode = mode,
+		size = size,
+		alignment = alignment,
+		old_memory = old_memory,
+		old_size = old_size,
+		location = loc,
+	)
+	duration := time.since(start_time)
+
+	data.modes[mode].count += 1
+	data.modes[mode].time += duration
+
+	return
+}
+
+
+// Allocator for testing control flow edge cases like allocation failures.
+// Note: uses global rand.
+Testing_Allocator :: struct {
+	backing:      Allocator,
+	error_chance: f32, // [0..1]
+}
+
+testing_allocator_init :: proc(t: ^Testing_Allocator, error_chance: f32 = 0.05, backing := context.allocator) {
+	t^ = {
+		backing = backing,
+		error_chance = error_chance,
+	}
+}
+
+@(require_results)
+testing_allocator :: proc(t: ^Testing_Allocator) -> Allocator {
+	return {
+		procedure = testing_allocator_proc,
+		data = t,
+	}
+}
+
+testing_allocator_proc :: proc(allocator_data: rawptr, mode: Allocator_Mode,
+                               size, alignment: int,
+                               old_memory: rawptr, old_size: int, loc := #caller_location) -> (result: []byte, err: Allocator_Error) {
+	data := cast(^Testing_Allocator)allocator_data
+
+	switch mode {
+	case .Alloc, .Alloc_Non_Zeroed, .Resize:
+		if rand.float32() < data.error_chance {
+			return nil, .Out_Of_Memory
+		}
+	case .Free, .Free_All, .Query_Features, .Query_Info:
+		// Ignore
+	}
+
+	return data.backing.procedure(
+		allocator_data = data.backing.data,
+		mode = mode,
+		size = size,
+		alignment = alignment,
+		old_memory = old_memory,
+		old_size = old_size,
+		location = loc,
+	)
+}
```

Comment:
Is this that useful in practice?

Reply:
I think stuff like this is used for large programs to test correctness, but I guess it's probably not useful enough to be in the `core` library. I will remove it from this PR.

Reply:
On the contrary, there is a `testing` package in `core` that could benefit from this.

Reply:
The ability to specify the "pass test" to user code is probably the better route.  The probabilistic route is neat because it can compress a huge set of possible tests as 1 test, which would complement a large state-space testing problem with decent coverage of explicit tests.

Reply:
Yeah, I guess an utility like this is pretty project specific. It also depends a lot on the error handling strategy, which might be different in each program.

Reply:
Also, I wasn't really thinking about 'tests' when writing this allocator. I feel like it would be much more useful when testing a large program, since that way you can check all of the obscure code paths which might otherwise never get executed.